fdbserver/MasterProxyServer.actor.cpp

```diff
@@ -1972,6 +1976,85 @@ ACTOR Future<Void> proxyCheckSafeExclusion(Reference<AsyncVar<ServerDBInfo>> db,
 	return Void();
 }
 
+ACTOR Future<Version> getTxnStateLifetime(Database db) {
+	state Transaction tr(db);
+	loop {
+		try {
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			Optional<Value> val = wait(tr.get(txStateLifetime));
+			if (val.present()) {
+				return BinaryReader::fromStringRef<Version>(val.get(), IncludeVersion());
+			} else {
+				tr.set(txStateLifetime, BinaryWriter::toValue(SERVER_KNOBS->TXN_STATE_LIFE_VERSIONS, IncludeVersion()));
+				wait(tr.commit());
+				return SERVER_KNOBS->TXN_STATE_LIFE_VERSIONS;
+			}
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+}
+
+ACTOR Future<std::deque<std::pair<Version, Version>>> readRecoveryVersions(Database db) {
+	state std::deque<std::pair<Version, Version>> result;
+	state Transaction tr(db);
+	loop {
+		try {
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			Standalone<RangeResultRef> range = wait(tr.getRange(recoveryVersions, CLIENT_KNOBS->TOO_MANY));
+			ASSERT(!range.more);
+			for (auto const& kv : range) {
+				auto from = BinaryReader::fromStringRef<Version>(kv.key.removePrefix(recoveryVersions.begin), Unversioned());
+				auto to = BinaryReader::fromStringRef<Version>(kv.value, IncludeVersion());
+				ASSERT(from < to);
+				ASSERT(result.empty() || result.back().second < from);
+				result.emplace_back(from, to);
+			}
+			return result;
+		} catch (Error& e) {
+			wait(tr.onError(e));
+		}
+	}
+}
+
+ACTOR Future<Void> txnStateCleaner(ProxyCommitData* self) {
+	state Transaction tr(self->cx);
+	state Version minVersion;
+	state KeyRange minVersionKey = singleKeyRange(transactionIDMinVersion);
+	state Version stateLifeTime = wait(getTxnStateLifetime(self->cx));
+	state std::deque<std::pair<Version, Version>> recoveryRanges = wait(readRecoveryVersions(self->cx));
+	stateLifeTime = std::min(stateLifeTime, SERVER_KNOBS->MAX_READ_TRANSACTION_LIFE_VERSIONS);
+	for (auto const& p : recoveryRanges) {
+		stateLifeTime += p.second - p.first;
+	}
+	loop {
+		wait(delay(SERVER_KNOBS->TXN_STATE_CLEANER_DELAY));
+		loop {
+			try {
+				tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+				Version currentVersion = wait(tr.getReadVersion());
+				minVersion = currentVersion - stateLifeTime;
+				if (minVersion > recoveryRanges.front().first) {
+					// we need to jump the recovery versions
+					minVersion -= recoveryRanges.front().second - recoveryRanges.front().first;
+					tr.clear(singleKeyRange(recoveryVersionsKey(recoveryRanges.front().first)));
+					recoveryRanges.pop_front();
+				}
+				auto clearTo = subkeyForTransactionVersion(minVersion);
+				KeyRangeRef toClean = KeyRangeRef(transactionIDs.begin, clearTo);
+				tr.addReadConflictRange(toClean);
```

Comment:
I think adding this read conflict range would mean that if something was happening with some transactions that caused them to repeatedly exceed the valid window, then the cleaner will fail with conflicts and might get starved.

Is it a necessary conflict? I'm not sure it's worth blocking this transaction if some delayed commit writes to the cleared space, but maybe there's some other consideration I haven't thought of.

Reply:
I think you're right - the conflict is not necessary. I don't think it can technically starve (as it is guaranteed that at least one transaction can commit) but it would make it very very slow...